### PR TITLE
[Build] allow to generate symbols for a subset of binaries

### PR DESCRIPTION
**Pull request by edymtt**
This would be needed to reduce overall build times in scenarios when
generating symbols for all binaries is too expensive and/or not needed.

At the same time, introduce tests around the logic that handles symbols.

Addresses rdar://76865276

### CODE REVIEW

**Comment by edymtt** [comment]
@swift-ci please smoke test

**Comment by edymtt** [comment]
@swift-ci Please Build Toolchain

**Comment by edymtt** [comment]
@swift-ci please smoke test macOS

**Comment by edymtt** [comment]
@swift-ci Please Build Toolchain macOS

**Comment by edymtt** [comment]
@swift-ci please smoke test macOS

File: validation-test/BuildSystem/extractsymbols-no-file-to-process.test [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# COM: Ensure the logic for extracting symbols does not fail is there is nothing to process
```
Comment (gottesmm) [inline validation-test/BuildSystem/extractsymbols-no-file-to-process.test:1]: Can you remove the COM: prefix. I think it will be clear that it is a comment.

File: utils/build-script-impl [hunk 2]
```diff
@@ -3158,16 +3185,11 @@ for host in "${ALL_HOSTS[@]}"; do
...
             #
-            # Exclude shell scripts and static archives.
-            # Exclude swift-api-digester dSYM to reduce debug toolchain size.
             # Tweak carefully the amount of parallelism -- dsymutil can be memory intensive and
             # as such too many instance can exhaust the memory and slow down/panic the machine
             printJSONStartTimestamp dsymutil
             (cd "${host_symroot}" &&
              find ./"${CURRENT_PREFIX}" -perm -0111 -type f -print | \
-               grep -v '.py$' | \
-               grep -v '.a$' | \
```
Comment (gottesmm) [inline utils/build-script-impl:3169]: This part of the change (moving the grep -v of the .py/.a/etc) seems to not be about generate symbols for a subset of binaries, can you do a git reset --soft and split this into a separate commit (same PR) that has a commit message that explains why this is being done. I am fine if you leave the test in the 2nd commit and just say in the first commit msg that you are putting the test in the 2nd commit.

File: utils/build-script-impl [hunk 3]
```diff
@@ -3130,15 +3160,10 @@ for host in "${ALL_HOSTS[@]}"; do
...
-        if [[ "${DRY_RUN}" ]]; then
-            call darwin_install_extract_symbols
-            printJSONStartTimestamp dsymutil
-            echo xargs -n 1 -P ${DSYMUTIL_JOBS} dsymutil
-            printJSONEndTimestamp dsymutil
-        else
+        # FIXME this if statement is a trick to have a more readable
+        # diff for the PR that made the following code
+        # amenable to dry run testing
+        if [ -n "${DRY_RUN}" -o -z "${DRY_RUN}" ]; then
```
Comment (gottesmm) [inline utils/build-script-impl:3166]: Can you eliminate this in a cleanup PR afterwards that also fixes the indentation. (Eric spoke with me and said that he was trying to reduce indentation changes, I am fine with that being done in a separate PR that is an easy rubber stamp).

File: utils/build-script-impl [hunk 4]
```diff
@@ -3147,7 +3172,9 @@ for host in "${ALL_HOSTS[@]}"; do
             # Copy executables and shared libraries from the `host_install_destdir` to
             # INSTALL_SYMROOT and run dsymutil on them.
             (cd "${CURRENT_INSTALL_DIR}" &&
-             find ./"${CURRENT_PREFIX}" -perm -0111 -type f -print | cpio --insecure -pdm "${host_symroot}")
+             find ./"${CURRENT_PREFIX}" -perm -0111 -type f -not -name "*.a" -not -name "*.py" -print | \
```
Comment (gottesmm) [inline utils/build-script-impl:3175]: Can you double check that we do not have any python in the sym root and any .a? My specific concern would be .py files from lldb.

File: validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+# COM: ensure we get all the values if we specify the flag multiple times
+# RUN: %empty-directory(%t/symroot)
+# RUN: mkdir -p %t/symroot/macosx-%target-cpu
+# RUN: SKIP_XCODE_VERSION_CHECK=1 SWIFT_BUILD_ROOT=%t %swift_src_root/utils/build-script --dry-run --cmake %cmake --darwin-install-extract-symbols=1 --install-destdir=%t/destdir --toolchain-prefix="" --install-symroot=%t/symroot --darwin-symroot-path-filters="/lib/" --darwin-symroot-path-filters="/swift-demangle" --jobs=1  2>&1 | tee %t/build-script-output.txt
+# RUN: %FileCheck --input-file=%t/build-script-output.txt %s
+# RUN: %FileCheck --input-file=%t/build-script-output.txt --check-prefixes CHECK-SKIPPED %s
+
+# REQUIRES: standalone_build
+# REQUIRES: OS=macosx
```
Comment (gottesmm) [inline validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test:36]: Can you move REQUIRES: to the top?

File: validation-test/BuildSystem/extractsymbols-default-behaviour.test [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+# RUN: chmod a+x %t/destdir/swift-util.py
+# RUN: %swiftc_driver %t/dylib.swift -emit-library -o %t/destdir/libswiftDemangle.dylib
+# RUN: %swiftc_driver %t/dylib.swift -emit-library -static -o %t/destdir/libswiftASTSectionImporter.a 
+# RUN: mkdir -p %t/symroot/macosx-%target-cpu
+
+# RUN: SWIFT_BUILD_ROOT=%t %swift_src_root/utils/build-script-impl --dry-run --build-dir=%t/build --workspace=%swift_src_root/.. --cmake %cmake --only-execute macosx-%target-cpu-extractsymbols --host-cc /usr/bin/true --darwin-install-extract-symbols=1 --host-target=macosx-%target-cpu --install-symroot=%t/symroot --install-destdir=%t/destdir --build-jobs=1 > %t/build-script-impl-output.txt 2>&1
+# RUN: %FileCheck --input-file=%t/build-script-impl-output.txt %s
+# RUN: %FileCheck --input-file=%t/build-script-impl-output.txt --check-prefixes CHECK-SKIPPED %s
+
+# REQUIRES: standalone_build
```
Comment (gottesmm) [inline validation-test/BuildSystem/extractsymbols-default-behaviour.test:18]: Same here, please move requires to the top.

**Review by gottesmm — COMMENTED** [review]
Looks good so far... can you make the changes/look into my questions. Then I want to do another quick go over/read the tests in a deeper way. Ping me to do another go around and assuming no further issues and the tests from first principles test everything we need, this will LGTM.

File: utils/build-script-impl (hunk 4 above)
Comment (edymtt) [reply utils/build-script-impl:3175]: We do have indeed python files in the symbols package

```
./usr/bin/swift-api-checker.py
./System/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/Python/lldb/utils/symbolication.py
./System/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/Python/lldb/utils/in_call_stack.py
./System/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/Python/lldb/macosx/crashlog.p
```

Comment (edymtt) [reply utils/build-script-impl:3175]: ...and also static libraries

```
./usr/lib/swift/appletvsimulator/libswiftCompatibility51.a
./usr/lib/swift/appletvsimulator/libswiftCompatibilityDynamicReplacements.a
./usr/lib/swift/appletvsimulator/libswiftCompatibility50.a
./usr/lib/swift/watchos/libswiftCompatibility51.a
./usr/lib/swift/watchos/libswiftCompatibilityDynamicReplacements.a
./usr/lib/swift/watchos/libswiftCompatibility50.a
./usr/lib/swift/iphoneos/libswiftCompatibility51.a
./usr/lib/swift/iphoneos/libswiftCompatibilityDynamicReplacements.a
./usr/lib/swift/iphoneos/libswiftCompatibility50.a
./usr/lib/swift/appletvos/libswiftCompatibility51.a
./usr/lib/swift/appletvos/libswiftCompatibilityDynamicReplacements.a
./usr/lib/swift/appletvos/libswiftCompatibility50.a
./usr/lib/swift/watchsimulator/libswiftCompatibility51.a
./usr/lib/swift/watchsimulator/libswiftCompatibilityDynamicReplacements.a
./usr/lib/swift/watchsimulator/libswiftCompatibility50.a
./usr/lib/swift/macosx/libswiftCompatibility51.a
./usr/lib/swift/macosx/libswiftCompatibilityDynamicReplacements.a
./usr/lib/swift/macosx/libswiftCompatibility50.a
./usr/lib/swift/iphonesimulator/libswiftCompatibility51.a
./usr/lib/swift/iphonesimulator/libswiftCompatibilityDynamicReplacements.a
./usr/lib/swift/iphonesimulator/libswiftCompatibility50.a
```

File: validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+# CHECK-DAG: codesign {{.*}}libswiftDemangle.dylib
+# CHECK-DAG: codesign {{.*}}lib_InternalSwiftScan.dylib
+
+# CHECK-SKIPPED: --- Extracting symbols ---
+
+# CHECK-SKIPPED-NOT: dsymutil {{.*}}swift-def-to-yaml-converter
+# CHECK-SKIPPED_NOT: dsymutil {{.*}}libswiftASTSectionImporter.a
+# CHECK-SKIPPED_NOT: dsymutil {{.*}}swift-util.py
+# CHECK-SKIPPED_NOT: dsymutil {{.*}}swift-api-digester
+# CHECK-SKIPPED_NOT: dsymutil {{.*}}lib_InternalSwiftScan.dylib
```
Comment (edymtt) [inline validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test:61]: There is a typo here -- it needs to be `-NOT` (hyphen NOT) and not `_NOT` (underscore NOT) for FileCheck to consider it

```suggestion
# CHECK-SKIPPED-NOT: dsymutil {{.*}}libswiftASTSectionImporter.a
# CHECK-SKIPPED-NOT: dsymutil {{.*}}swift-util.py
# CHECK-SKIPPED-NOT: dsymutil {{.*}}swift-api-digester
# CHECK-SKIPPED-NOT: dsymutil {{.*}}lib_InternalSwiftScan.dylib
```

Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test]: This is addressed in d3fb1c78b24f2c913dd59c0774fa1dfe1835608a

File: validation-test/BuildSystem/extractsymbols-no-file-to-process.test (hunk 1 above)
Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-no-file-to-process.test]: Those are now removed.

File: validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test (hunk 5 above)
Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test]: This is now addressed in d3fb1c78b24f2c913dd59c0774fa1dfe1835608a

File: validation-test/BuildSystem/extractsymbols-default-behaviour.test (hunk 6 above)
Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-default-behaviour.test]: This is addressed in fc5fa646715af4862655da49d87fcfb7fe6b36b3

File: utils/build-script-impl (hunk 4 above)
Comment (edymtt) [reply utils/build-script-impl]: In fc5fa646715af4862655da49d87fcfb7fe6b36b3 I kept this filter just before the `dsymutil` invocation -- I "folded" it into the `find` invocation so that we don't fail if no file is matched (as it could be the case with tests)

File: utils/build-script-impl (hunk 2 above)
Comment (edymtt) [reply utils/build-script-impl:3169]: After considering also the feedback in https://github.com/apple/swift/pull/37120#discussion_r626077819, I broke the PR in two commits as follows:

* fc5fa646715af4862655da49d87fcfb7fe6b36b3 focuses on getting this section under test and capturing the current behaviour (especially around populating the symroot when generating the toolchains)
* d3fb1c78b24f2c913dd59c0774fa1dfe1835608a introduce the filtering logic to copy only some paths to the symroot.

**Comment by edymtt** [comment]
@swift-ci please smoke test

**Comment by edymtt** [comment]
@swift-ci Please Build Toolchain

File: utils/build-script-impl (hunk 4 above)
Comment (edymtt) [reply utils/build-script-impl]: This is also explicitly tested in `extractsymbols-default-behaviour.test` -- to support this, `cpio` is now invoked with `-v` to print all copied paths.

**Comment by edymtt** [comment]
A quick check of the log shows that python files and compatibility libraries are copied to the symroot

```
> grep -A 100 cpio swift-PR-toolchain-osx-969.txt  | grep -e "\.a$" -e "\.py$"
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/bin/swift-api-checker.py
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/appletvsimulator/libswiftCompatibility51.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/appletvsimulator/libswiftCompatibilityDynamicReplacements.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/appletvsimulator/libswiftCompatibility50.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/watchos/libswiftCompatibility51.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/watchos/libswiftCompatibilityDynamicReplacements.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/watchos/libswiftCompatibility50.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/iphoneos/libswiftCompatibility51.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/iphoneos/libswiftCompatibilityDynamicReplacements.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/iphoneos/libswiftCompatibility50.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/appletvos/libswiftCompatibility51.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/appletvos/libswiftCompatibilityDynamicReplacements.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-x86_64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/appletvos/libswiftCompatibility50.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-arm64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/bin/swift-api-checker.py
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-arm64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/macosx/libswiftCompatibility51.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-arm64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/macosx/libswiftCompatibilityDynamicReplacements.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-arm64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/usr/lib/swift/macosx/libswiftCompatibility50.a
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-arm64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/System/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/Python/lldb/utils/symbolication.py
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-arm64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/System/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/Python/lldb/utils/in_call_stack.py
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-main/swift-nightly-symroot/macosx-arm64/.//Library/Developer/Toolchains/swift-PR-37120-969.xctoolchain/System/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/Python/lldb/macosx/crashlog.py
```

File: validation-test/BuildSystem/extractsymbols-no-file-to-process.test [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# REQUIRES: standalone_build
+# REQUIRES: OS=macosx
+
+# Ensure the logic for extracting symbols does not fail is there is nothing to process
```
Comment (gottesmm) [inline validation-test/BuildSystem/extractsymbols-no-file-to-process.test:4]: Can you add a better description of what this is actually testing? From our conversation it sounds like what this is really about is that the symbol extraction actually runs code even when we perform a dry-run and that previously this caused the dry-run code to fail (IIRC grep was specifically mentioned as the failing command). Can you expand the comment and add that context for when we both forget = p?

I would also change the name of this to extractsymbols-dry-run-does-not-fail.test or whatever works for you. Lets just go more for verbosity b/c people are not going to look at this code for a while.

File: validation-test/BuildSystem/extractsymbols-default-behaviour.test [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+# RUN: chmod a+x %t/destdir/libswiftCompatibility51.a
+# RUN: mkdir -p %t/symroot/macosx-%target-cpu
+
+# RUN: SWIFT_BUILD_ROOT=%t %swift_src_root/utils/build-script-impl --dry-run --build-dir=%t/build --workspace=%swift_src_root/.. --cmake %cmake --only-execute macosx-%target-cpu-extractsymbols --host-cc /usr/bin/true --darwin-install-extract-symbols=1 --host-target=macosx-%target-cpu --install-symroot=%t/symroot --install-destdir=%t/destdir --build-jobs=1 > %t/build-script-impl-output.txt 2>&1
+# RUN: %FileCheck --input-file=%t/build-script-impl-output.txt %s
+# RUN: %FileCheck --input-file=%t/build-script-impl-output.txt --check-prefixes CHECK-SKIPPED %s
+
+# CHECK: --- Extracting symbols ---
+
+# CHECK-LABEL: cpio
```
Comment (gottesmm) [inline validation-test/BuildSystem/extractsymbols-default-behaviour.test:26]: Can you add commentary to this file making it clear what each block is doing.

Example: this block could be labeled: "First check that we run cpio and copy files that we expect".

File: validation-test/BuildSystem/extractsymbols-default-behaviour.test [hunk 10]
```diff
@@ -0,0 +1,76 @@
+# REQUIRES: standalone_build
+# REQUIRES: OS=macosx
+
+# RUN: %empty-directory(%t)
+# RUN: mkdir -p %t
+# RUN: split-file %s %t
+
+# RUN: mkdir -p %t/destdir
```
Comment (gottesmm) [inline validation-test/BuildSystem/extractsymbols-default-behaviour.test:8]: Can you add a comment here explaining that:

1. Even though this is a test using dry-run, the symbol extraction code is not just echoed.
2. As a result, we want to make sure that the current behavior is at least tested and can take advantage of the (1) to generate a "simulated" build root and make sure that we do what we expect.

File: validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test [hunk 11]
```diff
@@ -0,0 +1,98 @@
+# REQUIRES: standalone_build
+# REQUIRES: OS=macosx
+
+# RUN: %empty-directory(%t)
+# RUN: mkdir -p %t
+# RUN: split-file %s %t
+
+# RUN: mkdir -p %t/destdir/bin
+# RUN: mkdir -p %t/destdir/lib
```
Comment (gottesmm) [inline validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test:9]: Can you also add the comment about the dry-run issue (as I mentioned on the other test) just to provide context for future readers.

**Review by gottesmm — APPROVED** [review]
LGTM. Can you address the comments about adding more context to the tests and then feel free to merge without coming back for further review!

Thank you for your work here!

**Comment by edymtt** [comment]
@swift-ci please smoke test

**Comment by edymtt** [comment]
@swift-ci Please Build Toolchain

File: validation-test/BuildSystem/extractsymbols-no-file-to-process.test (hunk 8 above)
Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-no-file-to-process.test]: Both points are addressed in b903a6cc658511ea333b357f05cbfd6362f17a17

File: validation-test/BuildSystem/extractsymbols-default-behaviour.test (hunk 9 above)
Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-default-behaviour.test:31]: Steps are now comments in b903a6cc658511ea333b357f05cbfd6362f17a17
(also reworked some `CHECK-LABEL` directives to ease readability)

File: validation-test/BuildSystem/extractsymbols-default-behaviour.test (hunk 10 above)
Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-default-behaviour.test:12]: Added comment in b903a6cc658511ea333b357f05cbfd6362f17a17

File: validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test (hunk 11 above)
Comment (edymtt) [reply validation-test/BuildSystem/extractsymbols-darwin-symroot-path-filters.test:13]: Added comments to file generations and steps in 1e5f790d06ea4d711daa73dbd9ee48d581be2163

File: utils/build-script-impl (hunk 3 above)
Comment (edymtt) [reply utils/build-script-impl:3165]: Sure, created follow up PR #37428
